tokio/src/net/addr.rs

```diff
@@ -40,6 +40,131 @@ where
     }
 }
 
+cfg_dns! {
+    use std::future::Future;
+    use std::pin::Pin;
+    use std::task::{Context, Poll};
+    /// LookupHost contains resolved SocketAddrs.
+    ///
+    /// This type is created by the function [`lookup_host`]. See its documentation for more.
+    #[derive(Debug)]
+    pub struct LookupHost<T>
+    where
+        T: sealed::ToSocketAddrsPriv,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        inner: LookupHostInner<T>,
+    }
+
+    impl<T> Future for LookupHost<T>
+    where
+        T: sealed::ToSocketAddrsPriv,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        type Output = io::Result<SocketAddr>;
+
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            if let LookupHostInner::Pending { fut } = &mut self.inner {
+                if let Poll::Ready(addrs) = Pin::new(fut).poll(cx) {
+                    match addrs {
+                        Ok(mut addrs) => {
+                            let addr = addrs.next().unwrap();
+                            Poll::Ready(Ok(addr))
+                        },
+                        Err(e) => Poll::Ready(Err(e))
+                    }
+                } else {
+                    Poll::Pending
+                }
+            } else {
+                unimplemented!()
+            }
+        }
+    }
+
+    // TODO: collapse this into a single enum once Rust 1.40 ships
+    // with the the `#[non_exhaustive]` attribute if Tokio's MSRV allows
+    //
+    // https://github.com/rust-lang/rust/pull/64639
+    #[derive(Debug)]
+    enum LookupHostInner<T>
+    where
+        T: sealed::ToSocketAddrsPriv,
+        T: ToSocketAddrs<Future = sealed::MaybeReady>,
+        T: ToSocketAddrs<Iter = sealed::OneOrMore>,
+    {
+        Pending {
+            fut: <T as sealed::ToSocketAddrsPriv>::Future,
+        },
+        Addrs {
+            addrs: sealed::OneOrMore,
```

Comment:
My guess is, to make it compile, this needs to be updated to: `<T as sealed::ToSocketAddrsPriv>::Iter`.

Reply:
I'm not really sure that's the case. That being said, what sort of signature _are_ you looking for here?